src/include/OpenImageIO/texture.h

```diff
@@ -1659,6 +1664,63 @@ class OIIO_API TextureSystem {
 
     /// @}
 
+    /// @{
+    /// @name Methods for UDIM patterns
+    ///
+
+    /// Is the filename a UDIM pattern?
+    ///
+    /// This method was added in OpenImageIO 2.3.
+    virtual bool is_udim(ustring filename) = 0;
+
+    /// Does the handle refer to a file that's a UDIM pattern?
+    ///
+    /// This method was added in OpenImageIO 2.3.
+    virtual bool is_udim(TextureHandle* udimfile) = 0;
+
+    /// For a UDIM filename pattern and texture coordinates, return the
+    /// TextureHandle pointer for the concrete tile file it refers to, or
+    /// nullptr if there is no corresponding tile (udim sets are allowed to
+    /// be sparse).
+    ///
+    /// This method was added in OpenImageIO 2.3.
+    virtual TextureHandle* resolve_udim(ustring udimpattern,
+                                        float s, float t) = 0;
+
+    /// A more efficient variety of `resolve_udim()` for cases where you
+    /// have the `TextureHandle*` that corresponds to the "virtual" UDIM
+    /// file and optionally have a `Perthread*` for the calling thread.
+    ///
+    /// This method was added in OpenImageIO 2.3.
+    virtual TextureHandle* resolve_udim(TextureHandle* udimfile,
+                                        Perthread* thread_info,
+                                        float s, float t) = 0;
+
+    /// Produce a full inventory of the set of concrete files comprising the
+    /// UDIM set specified by `udimpattern`.  The apparent number of texture
+    /// atlas tiles in the u and v directions will be written to `nutiles`
+    /// and `nvtiles`, respectively. The vector `filenames` will be sized
+    /// to `ntiles * nvtiles` and filled with the the names of the concrete
+    /// files comprising the atlas, with an empty ustring corresponding to
+    /// any unpopulated tiles (the UDIM set is allowed to be sparse). The
+    /// filename list is indexed as `utile + vtile * nvtiles`.
+    ///
+    /// This method was added in OpenImageIO 2.3.
+    virtual void inventory_udim(ustring udimpattern,
+                                std::vector<ustring>& filenames,
+                                int& nutiles, int& nvtiles) = 0;
```

Comment:
This looks appropriate for our current GPU preview shader. We use cuda's textures and the on demand library, so we assign consecutive ids to these tiles. At render time ids are computed from a pattern base_id uv's on the GPU. Requests for pixels are requested with these ids and the CPU then gets the filename and loads the data with the ImageCache.

For OSL we could query the shader for used patterns. Then map the pattern ustring hash to a base id and add the tile offset from uv's. This is because the peculiarities of the optix on demand library. I think we can work with this.

Don't know if there are plans for making oiio texture lookups work on GPU, for now we are reading pixels with oiio and sampling with cuda.

Reply:
I would like to work toward a future where OIIO migrates its TextureSystem to an "XPU" style where it will just work either from C++/CPU or from Cuda (minimizing how much code and data is redundantly/independently implemented for each path), and the Cuda side fully implements all current features of the TS. One step at a time.